src/crypto/merkle.rs

```diff
@@ -63,55 +83,78 @@ const EMPTY_ROOTS: [Hash; 24] = [
     hex!("b48a3973c4e0c5bad395ecd1c0cd430adc9680c7fd574460bf21c45f74840662"),
     hex!("826bfb27619fbf799f6000e4a3acdec264a13661e9c730e265c1c74c63501cb3"),
     hex!("720d56bf703cd7677805d33e73d2e7cb6104c932b62458ac3ae2f9d7469a1f38"),
+    hex!("ab2960c67fdd892b5eb1e63ac0c182c8d174671fd0f8065363895b5a3f7f0a88"),
+    hex!("ab7f4126b993ca5037a84190933bb8b97b9427a977d6872ca0b5d61470ea1834"),
+    hex!("dfaae89b54a08d497b2e8251de039fb38a7137b802d2d6c2554540c7b027721a"),
+    hex!("9451a516e349a64cd75fde5558bc92d3f359d3d49b560d2414b7eaa1e6889b1e"),
+    hex!("1c78453e429f9dbeaa70097b76abe42ad3da19a2ec5cda345aa77d713b930145"),
+    hex!("ee4878f4417f8dea52e54aaad49962a9fab5b746e54d6c49a5125e41e3ed6511"),
+    hex!("0cca41b299cd9be76a4b810fcbfff766e3f93ae2d8c7d761c0af2b524af97b56"),
+    hex!("33ebfb34d3aa119cb665d564acdd77b318dc86aa6a97744edb3bc03e97d776ca"),
 ];
-const_assert!(MAX_SLICES_PER_BLOCK.ilog2() < 24);
 
 /// Implementation of a Merkle tree.
 pub struct MerkleTree {
+    /// All hashes in the tree, leaf hashes and inner nodes.
     nodes: Vec<Hash>,
-    height: usize,
+    /// For each level, has the offset in `nodes` and the number of hashes on that level.
+    levels: Vec<(u32, u32)>,
 }
 
 impl MerkleTree {
     /// Creates a new Merkle tree from the given data for each leaf.
+    ///
+    /// This will always create a perfect binary tree (filling with empty leaves as necessary).
+    /// If you want to create a tree with more than half of the leaves empty,
+    /// you have to explicitly pass in empty leaves as part of `data`.
     pub fn new(data: &[impl AsRef<[u8]>]) -> Self {
         Self::new_from_iter(data.iter().map(|leaf| leaf.as_ref()))
     }
 
     /// Creates a new Merkle tree from the given data for each leaf.
+    ///
+    /// This will always create a perfect binary tree (filling with empty leaves as necessary).
+    /// If you want to create a tree with more than half of the leaves empty,
+    /// you have to explicitly pass in empty leaves as part of `data`.
     pub fn new_from_iter<'a>(data: impl IntoIterator<Item = &'a [u8]>) -> Self {
         let mut nodes = Vec::new();
-        let mut height = 0;
+        let mut levels = Vec::new();
 
         // calculate leaf hashes
         for leaf in data {
             let leaf_hash = hash_leaf(leaf);
             nodes.push(leaf_hash);
         }
-
-        // PERF: use `EMPTY_ROOTS` to get rid of these explicit empty leaves
-        let empty_leaf_hash = hash_leaf(&[]);
-        while !nodes.len().is_power_of_two() {
-            nodes.push(empty_leaf_hash);
-        }
+        levels.push((0, nodes.len() as u32));
```

Comment:
I highly doubt it'll ever happen but if there is an overflow, we will get weird behaviour.  


```suggestion
        levels.push((0, nodes.len().try_into().unwrap()));
```